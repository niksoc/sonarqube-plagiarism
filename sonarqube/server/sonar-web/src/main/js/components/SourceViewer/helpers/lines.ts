/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { intersection } from 'lodash';

export function optimizeHighlightedSymbols(
  symbolsForLine: string[] = [],
  highlightedSymbols: string[] = []
): string[] | undefined {
  const symbols = intersection(symbolsForLine, highlightedSymbols);

  return symbols.length ? symbols : undefined;
}

export function optimizeLocationMessage(
  highlightedLocationMessage: { index: number; text: string | undefined } | undefined,
  optimizedSecondaryIssueLocations: T.LinearIssueLocation[]
) {
  return highlightedLocationMessage != null &&
    optimizedSecondaryIssueLocations.some(
      location => location.index === highlightedLocationMessage.index
    )
    ? highlightedLocationMessage
    : undefined;
}

export function optimizeSelectedIssue(selectedIssue: string | undefined, issuesForLine: T.Issue[]) {
  return selectedIssue !== undefined && issuesForLine.find(issue => issue.key === selectedIssue)
    ? selectedIssue
    : undefined;
}
