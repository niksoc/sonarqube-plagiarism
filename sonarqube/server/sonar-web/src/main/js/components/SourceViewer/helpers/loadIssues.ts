/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { searchIssues } from '../../../api/issues';
import { getBranchLikeQuery } from '../../../helpers/branch-like';
import { parseIssueFromResponse } from '../../../helpers/issues';
import { BranchLike } from '../../../types/branch-like';

// maximum possible value
const PAGE_SIZE = 500;

function buildQuery(component: string, branchLike: BranchLike | undefined) {
  return {
    additionalFields: '_all',
    resolved: 'false',
    componentKeys: component,
    s: 'FILE_LINE',
    ...getBranchLikeQuery(branchLike)
  };
}

export function loadPage(
  query: T.RawQuery,
  page: number,
  pageSize = PAGE_SIZE
): Promise<T.Issue[]> {
  return searchIssues({
    ...query,
    p: page,
    ps: pageSize
  }).then(r =>
    r.issues.map(issue => parseIssueFromResponse(issue, r.components, r.users, r.rules))
  );
}

export function loadPageAndNext(
  query: T.RawQuery,
  toLine: number,
  page: number,
  pageSize = PAGE_SIZE
): Promise<T.Issue[]> {
  return loadPage(query, page).then(issues => {
    if (issues.length === 0) {
      return [];
    }

    const lastIssue = issues[issues.length - 1];

    if (
      (lastIssue.textRange != null && lastIssue.textRange.endLine > toLine) ||
      issues.length < pageSize
    ) {
      return issues;
    }

    return loadPageAndNext(query, toLine, page + 1, pageSize).then(nextIssues => {
      return [...issues, ...nextIssues];
    });
  });
}

export default function loadIssues(
  component: string,
  _fromLine: number,
  toLine: number,
  branchLike: BranchLike | undefined
): Promise<T.Issue[]> {
  const query = buildQuery(component, branchLike);
  return new Promise(resolve => {
    loadPageAndNext(query, toLine, 1).then(issues => {
      resolve(issues);
    });
  });
}
