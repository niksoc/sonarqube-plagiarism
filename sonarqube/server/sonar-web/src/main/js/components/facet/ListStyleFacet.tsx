/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import { sortBy, without } from 'lodash';
import * as React from 'react';
import ListFooter from 'sonar-ui-common/components/controls/ListFooter';
import SearchBox from 'sonar-ui-common/components/controls/SearchBox';
import Tooltip from 'sonar-ui-common/components/controls/Tooltip';
import { Alert } from 'sonar-ui-common/components/ui/Alert';
import { translate } from 'sonar-ui-common/helpers/l10n';
import { formatMeasure } from 'sonar-ui-common/helpers/measures';
import { queriesEqual } from 'sonar-ui-common/helpers/query';
import FacetBox from './FacetBox';
import FacetHeader from './FacetHeader';
import FacetItem from './FacetItem';
import FacetItemsList from './FacetItemsList';
import ListStyleFacetFooter from './ListStyleFacetFooter';
import MultipleSelectionHint from './MultipleSelectionHint';

interface SearchResponse<S> {
  maxResults?: boolean;
  results: S[];
  paging?: T.Paging;
}

export interface Props<S> {
  className?: string;
  disabled?: boolean;
  disabledHelper?: string;
  facetHeader: string;
  fetching: boolean;
  getFacetItemText: (item: string) => string;
  getSearchResultKey: (result: S) => string;
  getSearchResultText: (result: S) => string;
  loadSearchResultCount?: (result: S[]) => Promise<T.Dict<number>>;
  maxInitialItems: number;
  maxItems: number;
  minSearchLength: number;
  onChange: (changes: T.Dict<string | string[]>) => void;
  onClear?: () => void;
  onItemClick?: (itemValue: string, multiple: boolean) => void;
  onSearch: (query: string, page?: number) => Promise<SearchResponse<S>>;
  onToggle: (property: string) => void;
  open: boolean;
  property: string;
  query?: T.RawQuery;
  renderFacetItem: (item: string) => React.ReactNode;
  renderSearchResult: (result: S, query: string) => React.ReactNode;
  searchPlaceholder: string;
  getSortedItems?: () => string[];
  stats: T.Dict<number> | undefined;
  values: string[];
}

interface State<S> {
  autoFocus: boolean;
  query: string;
  searching: boolean;
  searchMaxResults?: boolean;
  searchPaging?: T.Paging;
  searchResults?: S[];
  searchResultsCounts: T.Dict<number>;
  showFullList: boolean;
}

export default class ListStyleFacet<S> extends React.Component<Props<S>, State<S>> {
  mounted = false;

  static defaultProps = {
    maxInitialItems: 15,
    maxItems: 100,
    minSearchLength: 2
  };

  state: State<S> = {
    autoFocus: false,
    query: '',
    searching: false,
    searchResultsCounts: {},
    showFullList: false
  };

  componentDidMount() {
    this.mounted = true;
  }

  componentDidUpdate(prevProps: Props<S>) {
    if (!prevProps.open && this.props.open) {
      // focus search field *only* if it was manually open
      this.setState({ autoFocus: true });
    } else if (
      (prevProps.open && !this.props.open) ||
      !queriesEqual(prevProps.query || {}, this.props.query || {})
    ) {
      // reset state when closing the facet, or when query changes
      this.setState({
        query: '',
        searchMaxResults: undefined,
        searchResults: undefined,
        searching: false,
        searchResultsCounts: {},
        showFullList: false
      });
    } else if (
      prevProps.stats !== this.props.stats &&
      Object.keys(this.props.stats || {}).length < this.props.maxInitialItems
    ) {
      // show limited list if `stats` changed and there are less than 15 items
      this.setState({ showFullList: false });
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleItemClick = (itemValue: string, multiple: boolean) => {
    if (this.props.onItemClick) {
      this.props.onItemClick(itemValue, multiple);
    } else {
      const { values } = this.props;
      if (multiple) {
        const newValue = sortBy(
          values.includes(itemValue) ? without(values, itemValue) : [...values, itemValue]
        );
        this.props.onChange({ [this.props.property]: newValue });
      } else {
        this.props.onChange({
          [this.props.property]: values.includes(itemValue) && values.length < 2 ? [] : [itemValue]
        });
      }
    }
  };

  handleHeaderClick = () => {
    this.props.onToggle(this.props.property);
  };

  handleClear = () => {
    if (this.props.onClear) {
      this.props.onClear();
    } else {
      this.props.onChange({ [this.props.property]: [] });
    }
  };

  stopSearching = () => {
    if (this.mounted) {
      this.setState({ searching: false });
    }
  };

  search = (query: string) => {
    if (query.length >= this.props.minSearchLength) {
      this.setState({ query, searching: true });
      this.props
        .onSearch(query)
        .then(this.loadCountsForSearchResults)
        .then(({ maxResults, paging, results, stats }) => {
          if (this.mounted) {
            this.setState(state => ({
              searching: false,
              searchMaxResults: maxResults,
              searchResults: results,
              searchPaging: paging,
              searchResultsCounts: { ...state.searchResultsCounts, ...stats }
            }));
          }
        })
        .catch(this.stopSearching);
    } else {
      this.setState({ query, searching: false, searchResults: [] });
    }
  };

  searchMore = () => {
    const { query, searchPaging, searchResults } = this.state;
    if (query && searchResults && searchPaging) {
      this.setState({ searching: true });
      this.props
        .onSearch(query, searchPaging.pageIndex + 1)
        .then(this.loadCountsForSearchResults)
        .then(({ paging, results, stats }) => {
          if (this.mounted) {
            this.setState(state => ({
              searching: false,
              searchResults: [...searchResults, ...results],
              searchPaging: paging,
              searchResultsCounts: { ...state.searchResultsCounts, ...stats }
            }));
          }
        })
        .catch(this.stopSearching);
    }
  };

  loadCountsForSearchResults = (response: SearchResponse<S>) => {
    const { loadSearchResultCount = () => Promise.resolve({}) } = this.props;
    const resultsToLoad = response.results.filter(result => {
      const key = this.props.getSearchResultKey(result);
      return this.getStat(key) === undefined && this.state.searchResultsCounts[key] === undefined;
    });
    if (resultsToLoad.length > 0) {
      return loadSearchResultCount(resultsToLoad).then(stats => ({ ...response, stats }));
    } else {
      return { ...response, stats: {} };
    }
  };

  getStat(item: string) {
    const { stats } = this.props;
    return stats && stats[item] !== undefined ? stats && stats[item] : undefined;
  }

  showFullList = () => {
    this.setState({ showFullList: true });
  };

  hideFullList = () => {
    this.setState({ showFullList: false });
  };

  renderList() {
    const { stats } = this.props;

    if (!stats) {
      return null;
    }

    const sortedItems = this.props.getSortedItems
      ? this.props.getSortedItems()
      : sortBy(
          Object.keys(stats),
          key => -stats[key],
          key => this.props.getFacetItemText(key)
        );

    const limitedList = this.state.showFullList
      ? sortedItems
      : sortedItems.slice(0, this.props.maxInitialItems);

    // make sure all selected items are displayed
    const selectedBelowLimit = this.state.showFullList
      ? []
      : sortedItems
          .slice(this.props.maxInitialItems)
          .filter(item => this.props.values.includes(item));

    const mightHaveMoreResults = sortedItems.length >= this.props.maxItems;

    return (
      <>
        <FacetItemsList>
          {limitedList.map(item => (
            <FacetItem
              active={this.props.values.includes(item)}
              key={item}
              name={this.props.renderFacetItem(item)}
              onClick={this.handleItemClick}
              stat={formatFacetStat(this.getStat(item))}
              tooltip={this.props.getFacetItemText(item)}
              value={item}
            />
          ))}
        </FacetItemsList>
        {selectedBelowLimit.length > 0 && (
          <>
            <div className="note spacer-bottom text-center">â‹¯</div>
            <FacetItemsList>
              {selectedBelowLimit.map(item => (
                <FacetItem
                  active={true}
                  key={item}
                  name={this.props.renderFacetItem(item)}
                  onClick={this.handleItemClick}
                  stat={formatFacetStat(this.getStat(item))}
                  tooltip={this.props.getFacetItemText(item)}
                  value={item}
                />
              ))}
            </FacetItemsList>
          </>
        )}
        <ListStyleFacetFooter
          count={limitedList.length + selectedBelowLimit.length}
          showLess={this.state.showFullList ? this.hideFullList : undefined}
          showMore={this.showFullList}
          total={sortedItems.length}
        />
        {mightHaveMoreResults && this.state.showFullList && (
          <Alert className="spacer-top" variant="warning">
            {translate('facet_might_have_more_results')}
          </Alert>
        )}
      </>
    );
  }

  renderSearch() {
    if (!this.props.stats || !Object.keys(this.props.stats).length) {
      return null;
    }

    return (
      <SearchBox
        autoFocus={this.state.autoFocus}
        className="little-spacer-top spacer-bottom"
        loading={this.state.searching}
        minLength={this.props.minSearchLength}
        onChange={this.search}
        placeholder={this.props.searchPlaceholder}
        value={this.state.query}
      />
    );
  }

  renderSearchResults() {
    const { searching, searchMaxResults, searchResults, searchPaging } = this.state;

    if (!searching && (!searchResults || !searchResults.length)) {
      return <div className="note spacer-bottom">{translate('no_results')}</div>;
    }

    if (!searchResults) {
      // initial search
      return null;
    }

    return (
      <>
        <FacetItemsList>
          {searchResults.map(result => this.renderSearchResult(result))}
        </FacetItemsList>
        {searchMaxResults && (
          <Alert className="spacer-top" variant="warning">
            {translate('facet_might_have_more_results')}
          </Alert>
        )}
        {searchPaging && (
          <ListFooter
            className="spacer-bottom"
            count={searchResults.length}
            loadMore={this.searchMore}
            ready={!searching}
            total={searchPaging.total}
          />
        )}
      </>
    );
  }

  renderSearchResult(result: S) {
    const key = this.props.getSearchResultKey(result);
    const active = this.props.values.includes(key);
    const stat = this.getStat(key) || this.state.searchResultsCounts[key];
    const disabled = !active && stat === 0;
    return (
      <FacetItem
        active={active}
        disabled={disabled}
        key={key}
        name={this.props.renderSearchResult(result, this.state.query)}
        onClick={this.handleItemClick}
        stat={formatFacetStat(stat)}
        tooltip={this.props.getSearchResultText(result)}
        value={key}
      />
    );
  }

  render() {
    const { disabled, stats = {} } = this.props;
    const { query, searching, searchResults } = this.state;
    const values = this.props.values.map(item => this.props.getFacetItemText(item));
    const loadingResults =
      query !== '' && searching && (searchResults === undefined || searchResults.length === 0);
    const showList = !query || loadingResults;
    return (
      <FacetBox
        className={classNames(this.props.className, {
          'search-navigator-facet-box-forbidden': disabled
        })}
        property={this.props.property}>
        <FacetHeader
          fetching={this.props.fetching}
          name={
            <Tooltip overlay={disabled ? this.props.disabledHelper : undefined}>
              <span>{this.props.facetHeader}</span>
            </Tooltip>
          }
          onClear={this.handleClear}
          onClick={disabled ? undefined : this.handleHeaderClick}
          open={this.props.open && !disabled}
          values={values}
        />

        {this.props.open && !disabled && (
          <>
            {this.renderSearch()}
            {showList ? this.renderList() : this.renderSearchResults()}
            <MultipleSelectionHint options={Object.keys(stats).length} values={values.length} />
          </>
        )}
      </FacetBox>
    );
  }
}

function formatFacetStat(stat: number | undefined) {
  return stat && formatMeasure(stat, 'SHORT_INT');
}
