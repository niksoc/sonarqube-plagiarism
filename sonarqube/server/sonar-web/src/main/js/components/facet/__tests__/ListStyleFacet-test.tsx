/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow, ShallowWrapper } from 'enzyme';
import * as React from 'react';
import { waitAndUpdate } from 'sonar-ui-common/helpers/testUtils';
import ListStyleFacet, { Props } from '../ListStyleFacet';

it('should render', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('should select items', () => {
  const onChange = jest.fn();
  const wrapper = shallowRender({ onChange });
  const instance = wrapper.instance() as ListStyleFacet<string>;

  // select one item
  instance.handleItemClick('b', false);
  expect(onChange).lastCalledWith({ foo: ['b'] });
  wrapper.setProps({ values: ['b'] });

  // select another item
  instance.handleItemClick('a', false);
  expect(onChange).lastCalledWith({ foo: ['a'] });
  wrapper.setProps({ values: ['a'] });

  // unselect item
  instance.handleItemClick('a', false);
  expect(onChange).lastCalledWith({ foo: [] });
  wrapper.setProps({ values: [] });

  // select multiple items
  wrapper.setProps({ values: ['b'] });
  instance.handleItemClick('c', true);
  expect(onChange).lastCalledWith({ foo: ['b', 'c'] });
  wrapper.setProps({ values: ['b', 'c'] });

  // unselect item
  instance.handleItemClick('c', true);
  expect(onChange).lastCalledWith({ foo: ['b'] });
});

it('should toggle', () => {
  const onToggle = jest.fn();
  const wrapper = shallowRender({ onToggle });
  wrapper.find('FacetHeader').prop<Function>('onClick')();
  expect(onToggle).toBeCalled();
});

it('should clear', () => {
  const onChange = jest.fn();
  const wrapper = shallowRender({ onChange, values: ['a'] });
  wrapper.find('FacetHeader').prop<Function>('onClear')();
  expect(onChange).toBeCalledWith({ foo: [] });
});

it('should search', async () => {
  const onSearch = jest.fn().mockResolvedValue({
    results: ['d', 'e'],
    paging: { pageIndex: 1, pageSize: 2, total: 3 }
  });
  const loadSearchResultCount = jest.fn().mockResolvedValue({ d: 7, e: 3 });
  const wrapper = shallowRender({ loadSearchResultCount, onSearch });

  // search
  wrapper.find('SearchBox').prop<Function>('onChange')('query');
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  expect(onSearch).lastCalledWith('query');
  expect(loadSearchResultCount).lastCalledWith(['d', 'e']);

  // load more results
  onSearch.mockResolvedValue({
    results: ['f'],
    paging: { pageIndex: 2, pageSize: 2, total: 3 }
  });
  loadSearchResultCount.mockResolvedValue({ f: 5 });
  wrapper.find('ListFooter').prop<Function>('loadMore')();
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  expect(onSearch).lastCalledWith('query', 2);

  // clear search
  onSearch.mockClear();
  loadSearchResultCount.mockClear();
  wrapper.find('SearchBox').prop<Function>('onChange')('');
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  expect(onSearch).not.toBeCalled();
  expect(loadSearchResultCount).not.toBeCalled();

  // search for no results
  onSearch.mockResolvedValue({ results: [], paging: { pageIndex: 1, pageSize: 2, total: 0 } });
  wrapper.find('SearchBox').prop<Function>('onChange')('blabla');
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  expect(onSearch).lastCalledWith('blabla');
  expect(loadSearchResultCount).not.toBeCalled();

  // search fails
  onSearch.mockRejectedValue(undefined);
  wrapper.find('SearchBox').prop<Function>('onChange')('blabla');
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot(); // should render previous results
  expect(onSearch).lastCalledWith('blabla');
  expect(loadSearchResultCount).not.toBeCalled();
});

it('should limit the number of items', () => {
  const wrapper = shallowRender({ maxInitialItems: 2, maxItems: 5 });
  expect(wrapper.find('FacetItem').length).toBe(2);

  wrapper.find('ListStyleFacetFooter').prop<Function>('showMore')();
  wrapper.update();
  expect(wrapper.find('FacetItem').length).toBe(3);

  wrapper.find('ListStyleFacetFooter').prop<Function>('showLess')();
  wrapper.update();
  expect(wrapper.find('FacetItem').length).toBe(2);
});

it('should show warning that there might be more results', () => {
  const wrapper = shallowRender({ maxInitialItems: 2, maxItems: 3 });
  wrapper.find('ListStyleFacetFooter').prop<Function>('showMore')();
  wrapper.update();
  expect(wrapper.find('Alert').exists()).toBe(true);
});

it('should reset state when closes', () => {
  const wrapper = shallowRender();
  wrapper.setState({
    query: 'foobar',
    searchResults: ['foo', 'bar'],
    searching: true,
    showFullList: true
  });

  wrapper.setProps({ open: false });
  checkInitialState(wrapper);
});

it('should reset search when query changes', () => {
  const wrapper = shallowRender({ query: { a: ['foo'] } });
  wrapper.setState({ query: 'foo', searchResults: ['foo'], searchResultsCounts: { foo: 3 } });
  wrapper.setProps({ query: { a: ['foo'], b: ['bar'] } });
  checkInitialState(wrapper);
});

it('should collapse list when new stats have few results', () => {
  const wrapper = shallowRender({ maxInitialItems: 2, maxItems: 3 });
  wrapper.setState({ showFullList: true });

  wrapper.setProps({ stats: { d: 1 } });
  expect(wrapper.state('showFullList')).toBe(false);
});

it('should display all selected items', () => {
  const wrapper = shallowRender({
    maxInitialItems: 2,
    stats: { a: 10, b: 5, c: 3 },
    values: ['a', 'b', 'c']
  });
  expect(wrapper).toMatchSnapshot();
});

it('should be disabled', () => {
  const wrapper = shallowRender({ disabled: true, disabledHelper: 'Disabled helper description' });
  expect(wrapper).toMatchSnapshot();
});

function shallowRender(props: Partial<Props<string>> = {}) {
  return shallow(
    <ListStyleFacet
      facetHeader="facet header"
      fetching={false}
      getFacetItemText={identity}
      getSearchResultKey={identity}
      getSearchResultText={identity}
      onChange={jest.fn()}
      onSearch={jest.fn()}
      onToggle={jest.fn()}
      open={true}
      property="foo"
      renderFacetItem={identity}
      renderSearchResult={identity}
      searchPlaceholder="search for foo..."
      stats={{ a: 10, b: 8, c: 1 }}
      values={[]}
      {...props}
    />
  );
}

function identity(str: string) {
  return str;
}

function checkInitialState(wrapper: ShallowWrapper) {
  expect(wrapper.state('query')).toBe('');
  expect(wrapper.state('searchResults')).toBe(undefined);
  expect(wrapper.state('searching')).toBe(false);
  expect(wrapper.state('searchResultsCounts')).toEqual({});
  expect(wrapper.state('showFullList')).toBe(false);
}
