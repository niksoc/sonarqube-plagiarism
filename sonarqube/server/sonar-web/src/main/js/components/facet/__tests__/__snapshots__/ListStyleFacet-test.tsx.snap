// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be disabled 1`] = `
<FacetBox
  className="search-navigator-facet-box-forbidden"
  property="foo"
>
  <FacetHeader
    fetching={false}
    name={
      <Tooltip
        overlay="Disabled helper description"
      >
        <span>
          facet header
        </span>
      </Tooltip>
    }
    onClear={[Function]}
    open={false}
    values={Array []}
  />
</FacetBox>
`;

exports[`should display all selected items 1`] = `
<FacetBox
  className=""
  property="foo"
>
  <FacetHeader
    fetching={false}
    name={
      <Tooltip>
        <span>
          facet header
        </span>
      </Tooltip>
    }
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={
      Array [
        "a",
        "b",
        "c",
      ]
    }
  />
  <SearchBox
    autoFocus={false}
    className="little-spacer-top spacer-bottom"
    loading={false}
    minLength={2}
    onChange={[Function]}
    placeholder="search for foo..."
    value=""
  />
  <FacetItemsList>
    <FacetItem
      active={true}
      disabled={false}
      halfWidth={false}
      key="a"
      loading={false}
      name="a"
      onClick={[Function]}
      stat="10"
      tooltip="a"
      value="a"
    />
    <FacetItem
      active={true}
      disabled={false}
      halfWidth={false}
      key="b"
      loading={false}
      name="b"
      onClick={[Function]}
      stat="5"
      tooltip="b"
      value="b"
    />
  </FacetItemsList>
  <div
    className="note spacer-bottom text-center"
  >
    â‹¯
  </div>
  <FacetItemsList>
    <FacetItem
      active={true}
      disabled={false}
      halfWidth={false}
      key="c"
      loading={false}
      name="c"
      onClick={[Function]}
      stat="3"
      tooltip="c"
      value="c"
    />
  </FacetItemsList>
  <ListStyleFacetFooter
    count={3}
    showMore={[Function]}
    total={3}
  />
  <MultipleSelectionHint
    options={3}
    values={3}
  />
</FacetBox>
`;

exports[`should render 1`] = `
<FacetBox
  className=""
  property="foo"
>
  <FacetHeader
    fetching={false}
    name={
      <Tooltip>
        <span>
          facet header
        </span>
      </Tooltip>
    }
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={Array []}
  />
  <SearchBox
    autoFocus={false}
    className="little-spacer-top spacer-bottom"
    loading={false}
    minLength={2}
    onChange={[Function]}
    placeholder="search for foo..."
    value=""
  />
  <FacetItemsList>
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="a"
      loading={false}
      name="a"
      onClick={[Function]}
      stat="10"
      tooltip="a"
      value="a"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="b"
      loading={false}
      name="b"
      onClick={[Function]}
      stat="8"
      tooltip="b"
      value="b"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="c"
      loading={false}
      name="c"
      onClick={[Function]}
      stat="1"
      tooltip="c"
      value="c"
    />
  </FacetItemsList>
  <ListStyleFacetFooter
    count={3}
    showMore={[Function]}
    total={3}
  />
  <MultipleSelectionHint
    options={3}
    values={0}
  />
</FacetBox>
`;

exports[`should search 1`] = `
<FacetBox
  className=""
  property="foo"
>
  <FacetHeader
    fetching={false}
    name={
      <Tooltip>
        <span>
          facet header
        </span>
      </Tooltip>
    }
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={Array []}
  />
  <SearchBox
    autoFocus={false}
    className="little-spacer-top spacer-bottom"
    loading={false}
    minLength={2}
    onChange={[Function]}
    placeholder="search for foo..."
    value="query"
  />
  <FacetItemsList>
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="d"
      loading={false}
      name="d"
      onClick={[Function]}
      stat="7"
      tooltip="d"
      value="d"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="e"
      loading={false}
      name="e"
      onClick={[Function]}
      stat="3"
      tooltip="e"
      value="e"
    />
  </FacetItemsList>
  <ListFooter
    className="spacer-bottom"
    count={2}
    loadMore={[Function]}
    ready={true}
    total={3}
  />
  <MultipleSelectionHint
    options={3}
    values={0}
  />
</FacetBox>
`;

exports[`should search 2`] = `
<FacetBox
  className=""
  property="foo"
>
  <FacetHeader
    fetching={false}
    name={
      <Tooltip>
        <span>
          facet header
        </span>
      </Tooltip>
    }
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={Array []}
  />
  <SearchBox
    autoFocus={false}
    className="little-spacer-top spacer-bottom"
    loading={false}
    minLength={2}
    onChange={[Function]}
    placeholder="search for foo..."
    value="query"
  />
  <FacetItemsList>
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="d"
      loading={false}
      name="d"
      onClick={[Function]}
      stat="7"
      tooltip="d"
      value="d"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="e"
      loading={false}
      name="e"
      onClick={[Function]}
      stat="3"
      tooltip="e"
      value="e"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="f"
      loading={false}
      name="f"
      onClick={[Function]}
      stat="5"
      tooltip="f"
      value="f"
    />
  </FacetItemsList>
  <ListFooter
    className="spacer-bottom"
    count={3}
    loadMore={[Function]}
    ready={true}
    total={3}
  />
  <MultipleSelectionHint
    options={3}
    values={0}
  />
</FacetBox>
`;

exports[`should search 3`] = `
<FacetBox
  className=""
  property="foo"
>
  <FacetHeader
    fetching={false}
    name={
      <Tooltip>
        <span>
          facet header
        </span>
      </Tooltip>
    }
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={Array []}
  />
  <SearchBox
    autoFocus={false}
    className="little-spacer-top spacer-bottom"
    loading={false}
    minLength={2}
    onChange={[Function]}
    placeholder="search for foo..."
    value=""
  />
  <FacetItemsList>
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="a"
      loading={false}
      name="a"
      onClick={[Function]}
      stat="10"
      tooltip="a"
      value="a"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="b"
      loading={false}
      name="b"
      onClick={[Function]}
      stat="8"
      tooltip="b"
      value="b"
    />
    <FacetItem
      active={false}
      disabled={false}
      halfWidth={false}
      key="c"
      loading={false}
      name="c"
      onClick={[Function]}
      stat="1"
      tooltip="c"
      value="c"
    />
  </FacetItemsList>
  <ListStyleFacetFooter
    count={3}
    showMore={[Function]}
    total={3}
  />
  <MultipleSelectionHint
    options={3}
    values={0}
  />
</FacetBox>
`;

exports[`should search 4`] = `
<FacetBox
  className=""
  property="foo"
>
  <FacetHeader
    fetching={false}
    name={
      <Tooltip>
        <span>
          facet header
        </span>
      </Tooltip>
    }
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={Array []}
  />
  <SearchBox
    autoFocus={false}
    className="little-spacer-top spacer-bottom"
    loading={false}
    minLength={2}
    onChange={[Function]}
    placeholder="search for foo..."
    value="blabla"
  />
  <div
    className="note spacer-bottom"
  >
    no_results
  </div>
  <MultipleSelectionHint
    options={3}
    values={0}
  />
</FacetBox>
`;

exports[`should search 5`] = `
<FacetBox
  className=""
  property="foo"
>
  <FacetHeader
    fetching={false}
    name={
      <Tooltip>
        <span>
          facet header
        </span>
      </Tooltip>
    }
    onClear={[Function]}
    onClick={[Function]}
    open={true}
    values={Array []}
  />
  <SearchBox
    autoFocus={false}
    className="little-spacer-top spacer-bottom"
    loading={false}
    minLength={2}
    onChange={[Function]}
    placeholder="search for foo..."
    value="blabla"
  />
  <div
    className="note spacer-bottom"
  >
    no_results
  </div>
  <MultipleSelectionHint
    options={3}
    values={0}
  />
</FacetBox>
`;
